File: Editorial/3001-3100/3070_Count_Submatrices_with_Top_Left_Element_and_Sum_Less_Than_k/TypeScript/1_Approach_2D_Prefix_Sum.ts
function countSubmatrices(grid: number[][], k: number): number {
    const m: number = grid.length;
    const n: number = grid[0].length;
    const cols: number[] = new Array(n).fill(0);
    let res: number = 0;

    for (let i = 0; i < m; i++) {
        let rows: number = 0;
        for (let j = 0; j < n; j++) {
            cols[j] += grid[i][j];
            rows += cols[j];
            if (rows <= k) {
                res++;
            }
        }
    }

    return res;
}

File: Editorial/3001-3100/3070_Count_Submatrices_with_Top_Left_Element_and_Sum_Less_Than_k/C/1_Approach_2D_Prefix_Sum.c
int countSubmatrices(int** grid, int gridSize, int* gridColSize, int k) {
    int m = gridSize;
    int n = *gridColSize;
    int* cols = (int*)calloc(n, sizeof(int));
    int res = 0;

    for (int i = 0; i < m; i++) {
        int rows = 0;
        for (int j = 0; j < n; j++) {
            cols[j] += grid[i][j];
            rows += cols[j];
            if (rows <= k) {
                res++;
            }
        }
    }

    free(cols);
    return res;
}
